File: app/src/main/java/hr/marin/weatherjava/presenters/WeatherPresenter.java
package hr.marin.weatherjava.presenters;

import android.view.View;
import android.widget.ImageView;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import hr.marin.weatherjava.models.ConsolidatedWeather;
import hr.marin.weatherjava.models.Location;
import hr.marin.weatherjava.services.WeatherService;
import hr.marin.weatherjava.services.listeners.IWeatherServiceListener;

public class WeatherPresenter {
    private View view;
    private WeatherService service;

    public WeatherPresenter(final View view) {
        this.view = view;
        this.service = new WeatherService();
    }

    public void getLocationWeather(int id) {
        view.showLoader();
        service.getLocationWeather(id, new IWeatherServiceListener<Location>() {
            @Override
            public void onSuccess(Location location) {
                view.onFetchSuccess(location);
                view.hideLoader();
            }

            @Override
            public void onFailure(Throwable t) {
                view.onError();
                view.hideLoader();
            }
        });
    }

    public interface View {
        void showLoader();
        void hideLoader();
        void onFetchSuccess(Location location);
        void onError();
    }
}

File: app/src/main/java/hr/marin/weatherkotlin/presenters/WeatherPresenter.kt
package hr.marin.weatherkotlin.presenters

import hr.marin.weatherkotlin.models.Location
import hr.marin.weatherkotlin.services.WeatherService
import hr.marin.weatherkotlin.services.listeners.IWeatherServiceListener

class WeatherPresenter(val view: View) {

    private val service: WeatherService = WeatherService()

    fun getLocationWeather(id: Int) {
        view.showLoader()
        service.getLocationWeather(id, object: IWeatherServiceListener<Location> {
            override fun onSuccess(response: Location) {
                view.onFetchSuccess(response)
                view.hideLoader()
            }

            override fun onFailure(t: Throwable) {
                view.onError()
                view.hideLoader()
            }
        })
    }

    interface View {
        fun showLoader()
        fun hideLoader()
        fun onFetchSuccess(location: Location)
        fun onError()
    }
}
